feat: Allow different RawData types

toseko opened the pull request.

killenb commented on include/FixedPointConverter.h:
```diff
@@ -331,87 +334,108 @@ namespace ChimeraTK {
+      try {
+        if(_isSigned) {
+          using converter_signed = boost::numeric::converter<int32_t, double,
+              boost::numeric::conversion_traits<int32_t, double>, boost::numeric::def_overflow_handler, Round<double>>;
```
This one has to convert to, argh., the signed type with the same number of bits as the RawType.

killenb commented on include/FixedPointConverter.h:
```diff
@@ -444,13 +468,83 @@ namespace ChimeraTK {
+    // compute masks of used and unused bits
+    // by using 1L (64 bit) this also works for 32 bits, which needs 33 bits during the calculation
+    // NOLINTNEXTLINE(hicpp-signed-bitwise)
+    _usedBitsMask = static_cast<RawType>((1L << nBits) - 1L);
```
This hack is gonna fail for 64 bit integers because we would need 65 a bit container.

killenb replied:
Yep, that looks good.

killenb commented on include/FixedPointConverter.h:
```diff
@@ -444,13 +468,83 @@ namespace ChimeraTK {
+    boost::fusion::for_each(_minCookedValues, initCoefficients(this));
+  }
 
+  using DEPRECATED_FIXEDPOINT_DEFAULT = int64_t;
```
Nopw, that would be int32_t

toseko commented on include/FixedPointConverter.h:
Yes, I agree. I think I even saw it, but somehow did not change.

toseko commented on include/FixedPointConverter.h:
```diff
@@ -444,13 +468,83 @@ namespace ChimeraTK {
+  /********************************************************************************************************************/
+  template<typename RawType>
+  const int FixedPointConverter<RawType>::zero = 0;
+
```
OK.

killenb commented on backends/LogicalNameMapping/src/LNMBitRangeAccessPlugin.cc:
```diff
@@ -159,7 +159,7 @@ namespace ChimeraTK::LNMBackend {
       }
 
-      auto value = fixedPointConverter.toRaw(buffer_2D[0][0]);
+      auto value = (uint32_t)fixedPointConverter.toRaw(buffer_2D[0][0]);
```
dito

killenb reviewed (commented):
maybe we should make a static assert for the moment that the FixedPointConverter is only instantiated with int32_t?

Apart from that I think we are ready to merge to the master. I would do a full final review af the remaining issues are fixed (only did the delta of the comments and the commit, and lost the overview)

toseko committed:
fix: Add test for 32 bit raw and fix for it.

killenb commented on include/FixedPointConverter.h:
```diff
@@ -241,81 +242,88 @@ namespace ChimeraTK {
           fpc.padUnusedBits(rawValue);
-          return numericToUserType<UserType>(*(reinterpret_cast<uint32_t*>(&rawValue)));
+          if constexpr(sizeof(RawType) == 4) {
+            return numericToUserType<UserType>(*(std::bit_cast<uint32_t*>(&rawValue)));
```
```std::bit_cast``` works on an object, not on a pointer (in contrast to the reinterpret_cast), at least according to the docu. I don't know if this does the right thing. Probably it does, because it changes one pointer type into another.

At least it's unnecessarily complicated. 
```
numericToUserType<UserType>(std::bit_cast<uint32_t>(rawValue));
```
should do the trick.

killenb replied:
I will move this to a different branch and directly resolve it based on the master

killenb commented on include/FixedPointConverter.h:
```diff
@@ -241,81 +242,88 @@ namespace ChimeraTK {
+          if constexpr(sizeof(RawType) == 4) {
+            return numericToUserType<UserType>(*(std::bit_cast<uint32_t*>(&rawValue)));
+          }
+          else if constexpr(sizeof(RawType) == 8) {
```
I am wondering if we should separate the 64 bit case into another commit. 

1. commit: Formally introduce RawType as template parameter, but have an assertion in the constructor that only allows int32_t as type.
2. commit: Test extension to 64 bits, then make it work.

We first get commit 1. ready and put it into the master before we are messing with too may things.

killenb commented on tests/executables_src/testFixedPointConverter.cpp:
```diff
@@ -25,6 +25,9 @@ using namespace boost::unit_test_framework;
 #define SIGNED_HEX_TO_DOUBLE(INPUT) static_cast<double>(static_cast<int32_t>(INPUT))
 #define SIGNED_HEX_TO_INT64(INPUT) static_cast<int64_t>(static_cast<int32_t>(INPUT))
 
+#define SIGNED_HEX16_TO_DOUBLE(INPUT) static_cast<double>(static_cast<int16_t>(INPUT))
```
What about SIGNED_HEX64_TOxxx ?

toseko committed:
feat: wip on testing the problem with convetion from -1 for signed, the same fractional bit as size of the converter

killenb commented on include/FixedPointConverter.h:
```diff
@@ -399,14 +406,32 @@ namespace ChimeraTK {
-          using converter_signed = boost::numeric::converter<int32_t, double,
-              boost::numeric::conversion_traits<int32_t, double>, boost::numeric::def_overflow_handler, Round<double>>;
-          raw = converter_signed::convert(d_cooked);
+          if constexpr(sizeof(RawType) == 4) {
```
We do not touch the code structure for now.

killenb commented on tests/executables_src/testFixedPointConverter.cpp:
```diff
@@ -904,34 +922,74 @@ BOOST_AUTO_TEST_CASE(testInt32_fraction32) {
+  checkToRaw(converter, -5.75, 0x80000000, "ToRaw8");
 
   checkToCooked(converter, 0x40000000, 0.25);
   checkToCooked(converter, 0xC0000000, -0.25);
```
Code seems to have resolved whatever this comment originally referred to.

mhier committed:
feat: add numeric converter

mhier committed:
fix: wrong include of own header

mhier committed:
chore: move Boolean and Void to separate headers

mhier committed:
chore: remove option to throw boost::numeric::bad_numeric_cast

This is not allowed according to the TransferElement specification.
Seems to be unused anyway.

mhier committed:
chore: replace boost::numeric::converter

d-rothe committed:
fix: MatchingMode=historized might use initial values

When the provided initial values were not yet consistent, but a data
update coming in turned them into a consistent set, the data made
available to the user was corrupt, due to unnecessary buffer switch happening

toseko committed:
feat: add tests for 64 bits width for converter

toseko commented:
This is old pull request. Part it consider was rewritten from scratch so it become obsolete.
